### ISSUE TITLE
Add unit tests for related entities on the generated app

### ISSUE THREAD

**Issue by ivanmonteiro** [jhipster/jhipster-dotnetcore#397]
##### **Overview of the feature request**

Add unit tests for related entities on the generated dotnet application.

As an example:
Unit tests for Employee should test adding and removing its Department relationship.

##### **Motivation for or Use Case**

Currently the unit tests only adds/removes the main entity.
Changes to the generator **may** cause issues regarding relationships and are not caught by tests.
This is important to make sure that new PR's are not breaking relationship functionality at this project.

##### **Related issues or PR**

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by kennethlong** [comment]
Does the Java template generate these test?  I was going to peek at the implementation if it does.

**Comment by ivanmonteiro** [comment]
> Does the Java template generate these test? I was going to peek at the implementation if it does.

Last time I checked the main Jhipster generator didn't generate tests for relationships. 
Currently the tests only check if relationship is required.

The idea would be to create unit tests that:
- creates the main entity 
- also creates the dependant entities 
- saves the entity graph to the database
- verifies that all entities were created/updated
- also deletes dependant entities and checks if it was properly deleted

**Comment by irwansyahwii** [comment]
@ivanmonteiro can I take this?

**Comment by ivanmonteiro** [comment]
> The idea would be to create unit tests that

Sure, if you have questions let me know

**Comment by irwansyahwii** [comment]
@ivanmonteiro I just reviewed the generated codes and wants to confirm several things:

The only entities generated are User, Role, and UserRole. So that means we need to create/update/deletion tests that will create/update/delete the User and its Roles. Is this correct?

**Comment by ivanmonteiro** [comment]
> @ivanmonteiro I just reviewed the generated codes and wants to confirm several things:
> 
> The only entities generated are User, Role, and UserRole. So that means we need to create/update/deletion tests that will create/update/delete the User and its Roles. Is this correct?

When we create entities, currently the generator creates <entityName>ControllerIntTest.cs tests:
https://github.com/jhipster/jhipster-dotnetcore/blob/e9d40ccec16c5ab4efea8645a73fa53a9d6a620a/generators/dotnetcore/templates/dotnetcore/test/Project.Test/Controllers/_persistClass_ControllerIntTest.cs.ejs#L206

It tests controllers actions like save/update/delete and check if the a child entity is null or not. It only checks if it is required.

But it does not create child entities and save together with the parent. Like if we have a Person entity that has an Address child entity, it does not create the Address child entity.

Ideally whe should have tests that both creates the parent AND child entity, and tests the relationship:

- a one-to-many relationship: we create the parend with a few child entities, save it, the delete one of child then check if it was persisted correctly.
- a one-to-one relationship: we create the parent with the child, try to modify a value in the child, save the parent, check if it was persisted correctly.

So the _persistClass_ControllerIntTest.cs.ejs template can be a starting point, but I have a suggestion: 

Keep it simple, if you can. 

The template file I mentioned (_persistClass_ControllerIntTest.cs.ejs) is more like an integration test (more complex). But this issue could probably be solved creating unit tests (simple), and could use the repository (or service classes) directly in the test instead of creating an httpClient, calling the controllers, etc...

**Comment by irwansyahwii** [comment]
@ivanmonteiro every time I tried to add a relationship to an entity it was always error:
```
================= Student =================
Fields
name (String) 

Relationships
lecturers (Lecturer) one-to-many



ERROR! An error occured while running jhipster-dotnetcore:entity#askForTableName
ERROR! ERROR! Cannot read properties of undefined (reading 'length')
TypeError: Cannot read properties of undefined (reading 'length')
    at default.askForTableName (file:///Users/irwansyah/Documents/PROJECTS/jhipster-dotnetcore/test-ride-entities/node_modules/generator-jhipster-dotnetcore/generators/entity/prompts.js:205:21)
    at default.executeTask (file:///Users/irwansyah/Documents/PROJECTS/jhipster-dotnetcore/test-ride-entities/node_modules/yeoman-generator/dist/actions/lifecycle.js:244:26)
    at env.queueTask.once (file:///Users/irwansyah/Documents/PROJECTS/jhipster-dotnetcore/test-ride-entities/node_modules/yeoman-generator/dist/actions/lifecycle.js:218:56)
    at runLoop.add.once (file:///Users/irwansyah/.nvm/versions/node/v20.15.1/lib/node_modules/generator-jhipster/node_modules/yeoman-environment/dist/environment-base.js:384:23)
    at Immediate.<anonymous> (/Users/irwansyah/.nvm/versions/node/v20.15.1/lib/node_modules/generator-jhipster/node_modules/grouped-queue/lib/subqueue.js:48:34)
    at process.processImmediate (node:internal/timers:478:21)
(base) irwansyah@Irwansyahs-MacBook-Pro test-ride-entities % 
```

This error happens in two scenarios:
1. I add a new entity and specified a relationship to non existence entity
2. I added two entities and when I tried to add a relationship to one of them

Should I fix the error first?

**Comment by Abu1982** [comment]
Hi, I'd like to work on this issue.

I've reviewed the thread and confirmed there is no active PR for it.

**Planned approach:** Analyze and fix based on error signals in jhipster/jhipster-dotnetcore

I'll submit a focused PR once the fix is ready and locally verified.

**Comment by Abu1982** [comment]
Hi, I'd like to work on this issue.

I've reviewed the thread and confirmed there is no active PR for it.

**Planned approach:** Analyze and fix based on error signals in jhipster/jhipster-dotnetcore

I'll submit a focused PR once the fix is ready and locally verified.
